tiledb/sm/misc/types.h

```diff
@@ -377,7 +380,149 @@ struct ChunkData {
   std::vector<DiskLayout> filtered_chunks_;
 };
 
-}  // namespace sm
-}  // namespace tiledb
+/**
+ * Typed range operations.
+ */
+template <typename T, typename Enable = T>
+struct RangeOperations {
+  /**
+   * Performs correctness checks for a valid range. If any validity checks fail
+   * an error status is returned.
+   *
+   * @param range The range to check validitity for.
+   * @return Status of the checks.
+   */
+  static Status check_is_valid_range(const Range& range);
+};
+
+template <typename T>
+struct RangeOperations<
+    T,
+    typename std::enable_if<std::is_integral<T>::value, T>::type> {
+  static Status check_is_valid_range(const Range& range) {
+    auto r = (const T*)range.data();
+    // Check range bounds
+    if (r[0] > r[1]) {
+      std::stringstream ss;
+      ss << "Lower range bound " << r[0]
+         << " cannot be larger than the higher bound " << r[1];
+      return Status_Error(ss.str());
+    }
+    // Return okay.
+    return Status::Ok();
+  };
+};
+
+template <typename T>
+struct RangeOperations<
+    T,
+    typename std::enable_if<std::is_floating_point<T>::value, T>::type> {
+  static Status check_is_valid_range(const Range& range) {
+    auto r = (const T*)range.data();
+    // Check for NaN
+    if (std::isnan(r[0]) || std::isnan(r[1])) {
+      return Status_Error("Range contains NaN");
+    }
+    if (std::isinf(r[0]) || std::isinf(r[1])) {
+      return Status_Error("Range contains Inf");
+    }
+    // Check range bounds
+    if (r[0] > r[1]) {
+      std::stringstream ss;
+      ss << "Lower range bound " << r[0]
+         << " cannot be larger than the higher bound " << r[1];
+      return Status::Ok();
+    }
+    // Return okay
+    return Status::Ok();
+  };
+};
+
+template <>
+struct RangeOperations<std::string, std::string> {
+  static Status check_is_valid_range(const Range&) {
+    return Status::Ok();
+  };
+};
```

Comment:
This matches the current implementation. Should there be a check that the starting value is less than the ending value?